singularity-eos/eos/eos_davis.hpp

```diff
@@ -12,18 +12,159 @@
+      const Real rho, const Real sie, Real *lambda = nullptr) const;
+  PORTABLE_INLINE_FUNCTION Real BulkModulusFromDensityTemperature(
+      const Real rho, const Real temperature, Real *lambda = nullptr) const;
+  PORTABLE_INLINE_FUNCTION Real BulkModulusFromDensityInternalEnergy(
+      const Real rho, const Real sie, Real *lambda = nullptr) const;
+  PORTABLE_INLINE_FUNCTION Real GruneisenParamFromDensityTemperature(
+      const Real rho, const Real temperature, Real *lambda = nullptr) const;
+  PORTABLE_INLINE_FUNCTION Real GruneisenParamFromDensityInternalEnergy(
+      const Real rho, const Real sie, Real *lambda = nullptr) const;
+  PORTABLE_INLINE_FUNCTION void FillEos(Real &rho, Real &temp, Real &energy, Real &press,
+                                        Real &cv, Real &bmod, const unsigned long output,
+                                        Real *lambda = nullptr) const;
+  PORTABLE_INLINE_FUNCTION
+  void ValuesAtReferenceState(Real &rho, Real &temp, Real &sie, Real &press, Real &cv,
+                              Real &bmod, Real &dpde, Real &dvdt,
+                              Real *lambda = nullptr) const;
+  PORTABLE_INLINE_FUNCTION void
+  DensityEnergyFromPressureTemperature(const Real press, const Real temp, Real *lambda,
+                                       Real &rho, Real &sie) const;
+  // Generic functions provided by the base class. These contain e.g. the vector
+  // overloads that use the scalar versions declared here
+  SG_ADD_BASE_CLASS_USINGS(DavisReactants)
+  PORTABLE_INLINE_FUNCTION
+  int nlambda() const noexcept { return 0; }
+  static constexpr unsigned long PreferredInput() { return _preferred_input; }
+  PORTABLE_INLINE_FUNCTION void PrintParams() const {
+    static constexpr char s1[]{"DavisReactants Params: "};
+    printf("%srho0:%e e0:%e P0:%e\nT0:%e A:%e B:%e\nC:%e G0:%e Z:%e\nalpha:%e "
+           "Cv0:%e\n",
+           s1, _rho0, _e0, _P0, _T0, _A, _B, _C, _G0, _Z, _alpha, _Cv0);
+  }
+  void inline Finalize() {}
+  static std::string EosType() { return std::string("DavisReactants"); }
+
+ private:
+  static constexpr Real onethird = 1.0 / 3.0;
+  Real _rho0, _e0, _P0, _T0, _A, _B, _C, _G0, _Z, _alpha, _Cv0;
+  // static constexpr const char _eos_type[] = "DavisReactants";
+  static constexpr unsigned long _preferred_input =
+      thermalqs::density | thermalqs::specific_internal_energy;
+  PORTABLE_INLINE_FUNCTION Real Ps(const Real rho) const;
+  PORTABLE_INLINE_FUNCTION Real Es(const Real rho) const;
+  PORTABLE_INLINE_FUNCTION Real Ts(const Real rho) const;
+  PORTABLE_INLINE_FUNCTION Real Gamma(const Real rho) const;
+  PORTABLE_INLINE_FUNCTION Real square(const Real x) const { return x * x; }
+  PORTABLE_INLINE_FUNCTION Real cube(const Real x) const { return x * x * x; }
+  PORTABLE_INLINE_FUNCTION Real fourth(const Real x) const { return x * x * x * x; }
+  PORTABLE_INLINE_FUNCTION Real fifth(const Real x) const { return x * x * x * x * x; }
```

Comment:
These should be `FORCEINLINE` and they are common to both Davis implementations, can we move them to be free functions or something like that to reduce duplication?

Reply:
I'll add the `FORCRINLINE`. 

Combining the implementations is a bit annoying---I moved them into the classes to avoid exposing them to global namespace. I can make a davis_utils` or `davis_impl` namespace to hide them again, it just involves peppering the code with `using namespace davis_utils` everywhere in all the functions. When I made the changes, I felt duplicating the functionality in both classes was the simpler option. I'm willing to be convinced otherwise.

Reply:
I created an issue: #170

Reply:
I mean, these aren't *davis* utilities; they are generic, it's just that nobody else needs them.

Reply:
@chadmeyer, these functions are reimplemented in starting in line 163

Reply:
I also think these should be implemented force inline and that they build on each other (cube = square*x, fourth=square*square, fifth = square*cube). This will produce faster code at all opt levels.

Reply:
This optimizes quite well

```c++
template<unsigned N>
constexpr auto sg_pow(const Real value)
{
  return value * sg_pow<N-1>(value);
}

template<>
constexpr auto sg_pow<0>(const Real value)
{
  return 1;
}
```

Reply:
I always enjoy a _good_ recursive function 😄

Reply:
I am looking at it in compiler explorer and it does not appear to optimize well, even at O3, only with O3 -ffast-math (or intel equivalent) will it optimize a fourth power to 2 multiplies. I wouldn't mind have a mode where we enable fast math, but I know it's suspect from a reproducibility standpoint.

Reply:
> appear to optimize well, even at O3, only with O3 -ffast-math (or intel equivalent) will it optimize a fourth power to 2 multiplie

Is this similar to what you see? https://godbolt.org/z/qzMbedE8T

Reply:
I modified slightly, you can see the individual function this way, and you'll see the recursive has 3 multiplies, the non-recursive has 2:
https://godbolt.org/z/5Tda4s5GY

Reply:
> I modified slightly, you can see the individual function this way, and you'll see the recursive has 3 multiplies, the non-recursive has 2: https://godbolt.org/z/5Tda4s5GY

After spending *way* too much time figuring this out, the issue here is getting the compiler do "reduction"; with `fast-math` enabled, the compiler would be allowed to "flatten" the recursive stack, and give the optimal sequence of multiplies.

The solution is interesting, but rather bloated (at least in my attempts) https://godbolt.org/z/roqdMxKc3 (with side-by-side comparison of two methods)

```c++
template <unsigned N, typename std::enable_if<N == 1>::type* = nullptr>
constexpr auto sg_pow(const Real& value) {
  return value;
}

template <unsigned N, typename std::enable_if<N % 2 == 0>::type* = nullptr>
constexpr auto sg_pow(const Real& value) {
  return sg_pow<N / 2>(value) * sg_pow<N / 2>(value);
}

template <unsigned N, typename std::enable_if<N % 2 != 0>::type* = nullptr>
constexpr auto sg_pow(const Real& value) {
  return value * sg_pow<N - 1>(value);
}
```

Reply:
This is neat, but I think it's probably overkill here. Agreed that we could move these simple utilities into a header in base. @mauneyc-LANL @dholladay00 what about just specializing the template to all the powers we actually use? That way we have a clean interface but don't have to fight with the compiler to get it to optimize.

Reply:
> This is neat, but I think it's probably overkill here. Agreed that we could move these simple utilities into a header in base. @mauneyc-LANL @dholladay00 what about just specializing the template to all the powers we actually use? That way we have a clean interface but don't have to fight with the compiler to get it to optimize.

what @dholladay00 suggested is easiest and correct, I just got carried away in godbolt

Reply:
Resolved as discussed in the standup today: Now `singularity::math_utils::pow<N>` with explicit specializations for the even expansions